### PR TITLE
IDAES Surrogates Best Practices Examples

### PR DESCRIPTION
**Pull request by bpaul4**
## Proposed changes:
- Add new set of example demonstrating IDAES surrogates sampling, training, validation, visualization, integration and optimization in IDAES flowsheets
- Set of four examples (ALAMO, PySMO, Keras, and surrogate selection), training script and associated files

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by lbianchi-lbl** [comment]
@bpaul4 I'd suggest to try the following:

- Since Alamo is not available on GitHub Actions, remove it from the source directory before the build step in `main.yml`. To do so, add something like this for both the `examples` job and `build-examples-docs` job (e.g. after L55 and again after L92):
     ```yaml
    - name: Skip notebooks that require Alamo
      run: |
        rm -r src/<path to the Alamo notebook>.ipynb    
     ```
- For Tensorflow/Keras, add the `tensorflow` dependency to `requirements.txt` (e.g. after the `pecos` on L16)
  - I'm still on the fence regarding whether it would be better to pin the version or not; for the sake of making our CI builds a bit simpler (or at least more predictable), I suggest to pin it to the version you're currently testing (i.e. the `requirements.txt` line should be `tensorflow==X.Y.Z` where `X.Y.Z` is whatever the version you get from e.g. `pip show tensorflow`)

File: requirements.txt
Comment (bpaul4) [reply requirements.txt:17]: I've only tested the notebooks for `tensorflow 2.8.0` since that is what I have installed locally. @lbianchi-lbl is there a specific (version 2.5.0, 2.7.0, ...) you would suggest using as a minimum version and could the version specified here be the cause of the failing build tests?

Comment (andrewlee94) [reply requirements.txt:17]: First and foremost, should we be specifying a version for tensorflow at all?

Comment (lbianchi-lbl) [reply requirements.txt:17]: @andrewlee94 pinning the Tensorflow version was my suggestion. My reasoning is that it makes our builds more reproducible, without any significant downside for users (since they would likely use `idaes get-examples` and not `requirements.txt` directly).

Comment (lbianchi-lbl) [inline requirements.txt:17]:
```suggestion
tensorflow==2.8.0
# this should have at least one blank line

```

**Review by lbianchi-lbl — DISMISSED** [review]
@bpaul4 I think the errors are caused by a missing newline at the end of `requirements.txt`. I've added a review comment for this.

File: .github/workflows/main.yml
Comment (lbianchi-lbl) [inline .github/workflows/main.yml:58]: The `-r` flag is only needed to remove directories, and is not needed when the argument is a file (or multiple files).

```suggestion
        rm src/Examples/SurrMod/FlowsheetOptimization/ALAMO_flowsheet_optimization.ipynb
```

File: .github/workflows/main.yml
Comment (lbianchi-lbl) [inline .github/workflows/main.yml:98]:
```suggestion
        rm src/Examples/SurrMod/FlowsheetOptimization/ALAMO_flowsheet_optimization.ipynb
```

**Comment by bpaul4** [comment]
Thank you @andrewlee94 and @lbianchi-lbl.

**Comment by OOAmusat** [comment]
@bpaul4 This looks really good. I have two initial comments/questions:

1. Is there a reason why you're going with the absolute rather than relative errors, especially given the scales being discussed?
2. For the best practices example, is there a better way to present the results than line plots? The lines might give the impression that the cases are connected when in fact they are not. Maybe dotplots or bar charts might be better options? @dangunter might be able to help with this. Just a thought.

**Comment by bpaul4** [comment]
@OOAmusat I chose to use absolute error and to include the data scale in the plot titles to reduce misrepresentation of the results. For example, relative or percent error on some of the mole fractions variables appears very high when the actual values are close to zero, although the absolute error is tiny. I'd appreciate suggestions on ways to better display the surrogate accuracy.

Regarding your second point, I'm happy to update the plots and would like to discuss some options. Additionally, viewing the notebooks on GitHub I see the training output is very long and I am wondering if I should push clean versions of the notebooks before we merge them.

**Comment by bpaul4** [comment]
@OOAmusat @dangunter would you be available to review these notebooks? Please see my responses to your comments above regarding results presentation.

**Review by OOAmusat — APPROVED** [review]
Looks good to me. The only thing I would mention is what you already highlighted: some of the outputs are fairly long and redirecting the displayed text to files might be a good option.

**Review by carldlaird — APPROVED** [review]
This looks reasonable to me. Great job, Brandon. Sometime we should take a look at the relative error in the Keras models - We should be able to reduce those.

**Pull request merged**
